drivers/clock_control/clock_control_nrf2_hsfll.c

```diff
@@ -0,0 +1,237 @@
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* TODO: add support for other HSFLLs */
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+#include <ld_dvfs_handler.h>
+
+#define FLAG_FREQ_CHANGE_CB_EXPECTED BIT(FLAGS_COMMON_BITS)
+
+#define NRFS_DVFS_TIMEOUT K_MSEC(CONFIG_CLOCK_CONTROL_NRF2_NRFS_DVFS_TIMEOUT_MS)
+
+/* Clock options sorted from lowest to highest frequency */
+static const struct clock_options {
+	uint32_t frequency;
+	enum dvfs_frequency_setting setting;
+} clock_options[] = {
+	{
+		.frequency = MHZ(64),
+		.setting = DVFS_FREQ_LOW,
+	},
+	{
+		.frequency = MHZ(128),
+		.setting = DVFS_FREQ_MEDLOW,
+	},
+	{
+		.frequency = MHZ(320),
+		.setting = DVFS_FREQ_HIGH,
+	},
+};
+
+struct hsfll_dev_data {
+	STRUCT_CLOCK_CONFIG(hsfll, ARRAY_SIZE(clock_options)) clk_cfg;
+	struct k_timer timer;
+};
+
+static void freq_setting_applied_cb(enum dvfs_frequency_setting new_setting)
+{
+	ARG_UNUSED(new_setting);
+
+	const struct device *dev = DEVICE_DT_INST_GET(0);
+	struct hsfll_dev_data *dev_data = dev->data;
+	atomic_val_t prev_flags;
+
+	/* Process only expected notifications (after sent requests). */
+	prev_flags = atomic_and(&dev_data->clk_cfg.flags,
+				~FLAG_FREQ_CHANGE_CB_EXPECTED);
+	if (prev_flags & FLAG_FREQ_CHANGE_CB_EXPECTED) {
+		k_timer_stop(&dev_data->timer);
+
+		clock_config_update_end(&dev_data->clk_cfg, 0);
+	}
+}
+
+static void hsfll_update_timeout_handler(struct k_timer *timer)
+{
+	struct hsfll_dev_data *dev_data =
+		CONTAINER_OF(timer, struct hsfll_dev_data, timer);
+
+	clock_config_update_end(&dev_data->clk_cfg, -ETIMEDOUT);
+}
+
+static void hsfll_work_handler(struct k_work *work)
+{
+	struct hsfll_dev_data *dev_data =
+		CONTAINER_OF(work, struct hsfll_dev_data, clk_cfg.work);
+	enum dvfs_frequency_setting required_setting;
+	uint8_t to_activate_idx;
+	int rc;
+
+	to_activate_idx = clock_config_update_begin(work);
+	required_setting = clock_options[to_activate_idx].setting;
+
+	/* Notify the DVFS service about the required setting. */
+	rc = dvfs_service_handler_change_freq_setting(required_setting);
+	if (rc < 0) {
+		clock_config_update_end(&dev_data->clk_cfg, rc);
+		return;
+	}
+
+	/* And expect a confirmation that the setting is ready to be used. */
+	(void)atomic_or(&dev_data->clk_cfg.flags, FLAG_FREQ_CHANGE_CB_EXPECTED);
+	k_timer_start(&dev_data->timer, NRFS_DVFS_TIMEOUT, K_NO_WAIT);
+}
+
+static struct onoff_manager *hsfll_find_mgr(const struct device *dev,
+					    const struct nrf_clock_spec *spec)
+{
+	struct hsfll_dev_data *dev_data = dev->data;
+	uint32_t frequency;
+
+	if (!spec) {
+		return &dev_data->clk_cfg.onoff[0].mgr;
+	}
+
+	if (spec->accuracy || spec->precision) {
+		LOG_ERR("invalid specification of accuracy or precision");
+		return NULL;
+	}
+
+	frequency = spec->frequency == NRF_CLOCK_CONTROL_FREQUENCY_MAX
+		  ? MHZ(320)
```

Comment:
This value is quite hidden here. It would be better to have it defined somewhere near the `clock_options` array.

Reply:
added defines
```
#define HSFLL_FREQ_LOW    MHZ(64)
#define HSFLL_FREQ_MEDLOW MHZ(128)
#define HSFLL_FREQ_HIGH   MHZ(320)
```
and used this across file